backend/common/utils/rollup.py

```diff
@@ -55,30 +105,36 @@ def find_descendants_per_cell_type(cell_types):
     descendants_per_cell_type : list
         List of lists of descendants for each cell type in the input list.
     """
-    # a lookup table to avoid redundant ontology lookups
     lookup_table = {}
-
-    # a set of all cell types in the input list
-    # used to compute intersections with descendants
-    cell_types_set = set(cell_types)
-    relatives_per_cell_type = []
+    cell_types_set = frozenset(cell_types)
+    valid_descendants = []
     for cell_type in cell_types:
-        # if the input cell types are suffixed, only consider descendants within the same group
-        if ";;" in cell_type:
-            prefix, suffix = cell_type.split(";;")
-            suffix = ";;" + suffix
-        else:
-            prefix = cell_type
-            suffix = ""
         if cell_type not in lookup_table:
-            # find descendants of cell type and re-suffix them
-            relatives = [f"{i}{suffix}" for i in descendants(prefix)]
-            # only consider descendants (+suffixes) that are in the input list
-            lookup_table[cell_type] = list(cell_types_set.intersection(relatives))
+            lookup_table[cell_type] = get_valid_descendants(cell_type, cell_types_set)
+        valid_descendants.append(lookup_table[cell_type])
+    return valid_descendants
 
-        relatives_per_cell_type.append(lookup_table[cell_type])
 
-    return relatives_per_cell_type
+def are_cell_types_not_redundant_nodes(cell_types, cell_counts):
+    """
+    Determines whether each cell type in a list of cell types is not a redundant node.
+
```

Comment:
You should say a bit about the heuristic because it is not obvious at all. Something like:

If a cell-type `A` has the same cell-count as its descendant cell-type `B`, then we conclude that `A` and `B` must be connected by a single relationship chain where `A` and the cell-types leading `A` to `B` in the chain have zero cell-count and therefore the entire chain `[A,B)` (that is excluding `B`) will be excluded.

And then refer the reader to `get_valid_descendants()` to understand how the exclusion is done